include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,712 @@
+     * User's Guide.
+     */
+    template <typename T>
+    T
+    attr(const std::string attr_name);
+
+    /**
+     * Writes an attribute. T can be float, double, std::complex<float>,
+     * std::complex<double>, int, unsigned int, bool or std::string. Note that
+     * the encoding of std::string is UTF8 in order to be compatible with
+     * python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    void
+    write_attr(const std::string attr_name, const T value);
+
+
+    const std::string name;
+
+  protected:
+    /**
+     * HDF5 identifier for the objects File, Group and DataSet
+     */
+    std::shared_ptr<hid_t> hdf5_reference;
+
+    /**
+     * If true use parallel HDF5, if false use serial HDF5.
+     */
+    const bool mpi;
+
+    /**
+     * Returned value of the last HDF5 C library call.
+     *
+     * The HDF5 calls return a non-negative value if successful; otherwise they
+     * return a negative value.
+     *
+     * Calls such as
+     * [H5DWrite](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5D.html#Dataset-Write)
+     * use this variable to store the returned value. Other calls such as
+     * [H5Pcreate](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-Create)
+     * don't use this variable to store the returned value.
+     */
+    herr_t ret;
+  };
+
+  /**
+   * This class implements a HDF5 DataSet.
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class DataSet : public HDF5Object
+  {
+    friend class Group;
+
+  protected:
+    // Open dataset
+    DataSet(const std::string name, const hid_t &parent_group_id, bool mpi);
+
+    // Create dataset
+    DataSet(const std::string      name,
+            const hid_t &          parent_group_id,
+            std::vector<hsize_t>   dimensions,
+            std::shared_ptr<hid_t> t_type,
+            bool                   mpi);
+
+  public:
+    /**
+     * Reads data of the dataset. Number can be float, double,
+     * std::complex<float>, std::complex<double>, int or unsigned int.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <template <class...> class Container, typename number>
+    Container<number>
+    read();
```

Comment:
That makes sense! I didn't think about the `typename Container::value_type` option :) That is a better option. I followed your advice, although I have to admit that I like the concept of template template

I added in several places of the documentation:
`Container` can be `std::vector<float>`, `std::vector<double>`, `std::vector<std::complex<float>>`, `std::vector<std::complex<double>>`, `std::vector<int>`, `std::vector<unsigned int>`, `Vector<float>`, `Vector<double>`, `Vector<std::complex<float>>`, `Vector<std::complex<double>>`, `FullMatrix<float>`, `FullMatrix<double>`, `FullMatrix<std::complex<float>>` or `FullMatrix<std::complex<double>>`

Reply:
Nice, thank you!

The problem with template template arguments, as elegant as they are, is that you restrict yourself to objects that can be created with only a single argument because your return type is `Container<number>`. But that means that you can't, for example, use a `std::vector` with a custom allocator, or a `Table<2,double>`, etc. I know that you may not be interested in using such types in the current context, but template template arguments are in general pretty restrictive.

Reply:
That is a good point, actually a type such as `Table<4, double>` would be interesting! I was planing to add support in a future PR for `Tensor`, and now I see that `Table<N, number>` is also a good candidate to represent high rank datasets.  The size of Tensor is limited by the dimensions of the space, indices go from 0 to dim. Whereas there is no limit for Table. I'll explore this in the future